### DISCUSSION TITLE
Working with Volar

### DISCUSSION THREAD

**Discussion by johnsoncodehk** [vuejs/vetur#2842] [General]
[Volar](https://github.com/johnsoncodehk/volar) is a Vue 3 language support extension built for resolve [TS performance issue](https://github.com/vuejs/vetur/issues/1051). It's a subset of Vetur. Not include some features like Snippet, ESLint and Prettier.

Some Vetur and Volar duplicate features can't not working together (like Renaming), and may cause performance issues.

I believe if exclude duplicate features, this two extensions could working together as well. Can we disable LSP features if detect user installed and enabled Volar? This is a big UX / DX improvement for users. users don't need to disable one in Vue 3 project, and got complete DX experience.

If Vetur is willing to implement, I recommend add option for that and default enable, but show a message box when trigger. Because it will more out of the box, and some users even don't know this problem.

**Comment by yoyo930021** [comment]
I think we resolve 90% [TS performance issue](https://github.com/vuejs/vetur/issues/1051) in this project. 🤣

> I believe if exclude duplicate features, this two extensions could working together as well. Can we disable LSP features if detect user installed and enabled Volar? This is a big UX / DX improvement for users. users don't need to disable one in Vue 3 project, and got complete DX experience.

The Snippet, ESLint, and Prettier features are in the LSP.
So we can't quickly combine together.
@octref Do you have any idea?

I think we can have more cooperation with other projects.
We can focus more on our respective areas of expertise.
List the people who may be relevant:
Vetur: @octref @yoyo930021 @ktsn 
VueDX: @znck 
Volar: @johnsoncodehk

**Reply by znck** [reply]
@johnsoncodehk I haven't look into details but it seems VueDX and Volar have many similarities. I strongly feel that we should combine the efforts and try to minimise the duplicate work. 

I'll spend some time to understand internals of Volar. (Please share if there's a design document.)

In terms of unification, I believe vetur should ingest features from volar and VueDX. (We should schedule a call to discuss in detail.)

**Reply by johnsoncodehk** [reply]
> I'll spend some time to understand internals of Volar. (Please share if there's a design document.)

I am still refactoring, the code will have a lot of changes, I suggest you do it later.

> I strongly feel that we should combine the efforts and try to minimise the duplicate work.

Actually I think duplicate work is not a problem, sometimes independent work can develop faster.

We should pay attention to the final DX and ecology, and see where we can work together on this basis (such as syntax highlighting, e2e tests), otherwise I think it is difficult for DX to make progress in the past.

> In terms of unification, I believe vetur should ingest features from volar and VueDX.

I believe Vetur and VueDX can cooperate very well, but it conflicts with some of Volar's design and implement, because Volar intentionally reduces some things. 

> We should schedule a call to discuss in detail.

I really sorry, I can't talk with English 😅, but text is ok.

**Reply by johnsoncodehk** [reply]
Btw this topic more aims to solve the UX problem, currently in Vue 2 user need to disable Volar, in Vue 3 disable Vetur.

The ideal situation of Vetur is user never need to disable Vetur, but only need to disable Volar in Vue 2.

**Reply by yoyo930021** [reply]
More detail waiting for team discuss in Vue team.
